server/lib/src/valueset/eckey.rs

```diff
@@ -0,0 +1,220 @@
+        false
+    }
+
+    fn len(&self) -> usize {
+        1
+    }
+
+    fn generate_idx_eq_keys(&self) -> Vec<String> {
+        Vec::with_capacity(0)
+    }
+
+    fn syntax(&self) -> SyntaxType {
+        SyntaxType::EcKeyPrivate
+    }
+
+    fn validate(&self, _schema_attr: &crate::schema::SchemaAttribute) -> bool {
+        match self.set.as_ref() {
+            Some(key) => key.priv_key.check_key().is_ok() && key.pub_key.check_key().is_ok(),
+            None => false,
+        }
+    }
+
+    fn to_proto_string_clone_iter(&self) -> Box<dyn Iterator<Item = String> + '_> {
+        Box::new(iter::once(String::from("hidden")))
+    }
+
+    fn to_db_valueset_v2(&self) -> DbValueSetV2 {
+        #[allow(clippy::expect_used)]
+        let key_der = self
+            .set
+            .as_ref()
+            .map(|key| {
+                key.priv_key.private_key_to_der().expect(
+        "Unable to process eckey to der, likely corrupted. You must restore from backup.",
+    )
+            })
+            .unwrap_or_default();
+        DbValueSetV2::EcKeyPrivate(key_der)
+    }
+
+    fn to_repl_v1(&self) -> ReplAttrV1 {
+        #[allow(clippy::expect_used)]
+        let key_der = self
+            .set
+            .as_ref()
+            .map(|key| {
+                key.priv_key.private_key_to_der().expect(
+        "Unable to process eckey to der, likely corrupted. You must restore from backup.",
+    )
+            })
+            .unwrap_or_default();
+        ReplAttrV1::EcKeyPrivate { key: key_der }
+    }
+
+    fn to_partialvalue_iter(&self) -> Box<dyn Iterator<Item = crate::value::PartialValue> + '_> {
+        Box::new(iter::once(PartialValue::SecretValue))
+    }
+
+    fn to_value_iter(&self) -> Box<dyn Iterator<Item = crate::value::Value> + '_> {
+        match &self.set {
+            Some(key) => Box::new(iter::once(Value::EcKeyPrivate(key.priv_key.clone()))),
+            None => Box::new(iter::empty::<Value>()),
+        }
+    }
+
+    fn equal(&self, other: &super::ValueSet) -> bool {
+        #[allow(clippy::expect_used)]
+        other.as_ec_key_private().map_or(false, |other_key| {
+            self.set.as_ref().map_or(false, |key| {
+                key.priv_key
+                    .private_key_to_der()
+                    .expect("Failed to retrieve key der")
+                    == other_key
+                        .private_key_to_der()
+                        .expect("Failed to retrieve key der")
+            })
+        })
+    }
+
+    fn merge(&mut self, other: &super::ValueSet) -> Result<(), kanidm_proto::v1::OperationError> {
+        if let Some(other_key) = other.as_ec_key_private() {
+            let priv_key = other_key.clone();
+            #[allow(clippy::expect_used)]
+            let pub_key = Self::private_key_to_public_key(&priv_key).expect(
+                "Unable to retrieve public key from private key, likely corrupted. You must restore from backup.",
```

Comment:
This should log an error-level message and return an operation error, not panic.

Reply:
I agree but currently the ValueSet trait expects a direct result, so I went for the same approach as in `Jwt` for example.
There we have something like
```
 fn to_repl_v1(&self) -> ReplAttrV1 {
        ReplAttrV1::JwsKeyEs256 { set: self.set.iter()
            .map(|k| {
                #[allow(clippy::expect_used)]
                k.private_key_to_der()
                    .expect("Unable to process private key to der, likely corrupted. You must restore from backup.")
            })
            .map(|b| b.into())
            .collect()
        }
    }
```
I could open an issue for this tho

Reply:
but this one has a `Result<>` response, which is why I mentioned it 😄 

```rust
fn merge(&mut self, other: &super::ValueSet) -> Result<(), kanidm_proto::v1::OperationError> {
```

Reply:
Yep, it's a bummer that some do and some other don't. I wonder if soon we'll have to pay the tech debt again on this one :p

Reply:
If I remember correctly William also mentioned that the valueset trait could be redesigned, since it's based on the assumption that every value will be stored in a set, and that's not always true, as it's the case for the EcdhKey.
If this ever happens it could also fix the return values of the trait functions.

Reply:
Yeah, I think it needs a redesign with schema to have the multivalue type removed from schema, and value being single/multi is a property of the type instead.

Reply:
Either way, less panics, more errors in the meantime please 😃 Then this one's ready to go!

Reply:
Wopsies I must be blind I thought you wanted me to remove the expect on the method that didn't return the result. 
Anyway I fixed it!